sys/pm_layered/pm.c

```diff
@@ -36,52 +36,52 @@
 /**
  * @brief Global variable for keeping track of blocked modes
  */
-volatile pm_blocker_t pm_blocker = { .val_u32 = PM_BLOCKER_INITIAL };
+static pm_blocker_t pm_blocker = { .val_u32 = PM_BLOCKER_INITIAL };
 
 void pm_set_lowest(void)
 {
-    pm_blocker_t blocker = pm_blocker;
+    unsigned state = irq_disable();
     unsigned mode = PM_NUM_MODES;
     while (mode) {
-        if (blocker.val_u8[mode-1]) {
+        if (pm_blocker.val_u8[mode-1]) {
             break;
         }
         mode--;
     }
 
-    /* set lowest mode if blocker is still the same */
-    unsigned state = irq_disable();
-    if (blocker.val_u32 == pm_blocker.val_u32) {
-        DEBUG("pm: setting mode %u\n", mode);
-        pm_set(mode);
-    }
-    else {
-        DEBUG("pm: mode block changed\n");
-    }
+    pm_set(mode);
     irq_restore(state);
 }
 
 void pm_block(unsigned mode)
 {
-    assert(pm_blocker.val_u8[mode] != 255);
-
     unsigned state = irq_disable();
+    assert(pm_blocker.val_u8[mode] != 255);
     pm_blocker.val_u8[mode]++;
     irq_restore(state);
 }
 
 void pm_unblock(unsigned mode)
 {
+    unsigned state = irq_disable();
     assert(pm_blocker.val_u8[mode] > 0);
+    pm_blocker.val_u8[mode]--;
+    irq_restore(state);
+}
 
+pm_blocker_t pm_get_blocker(void)
+{
+    pm_blocker_t result;
     unsigned state = irq_disable();
-    pm_blocker.val_u8[mode]--;
+    result = pm_blocker;
     irq_restore(state);
+    return result;
 }
 
 #ifndef PROVIDES_PM_LAYERED_OFF
 void pm_off(void)
 {
+    irq_disable();
```

Comment:
I understand that disabling IRQ makes sure we sleep, but still doesnt feel right to add this change.

Reply:
Huh? The function is called `pm_off()`. We should be sure that after this call nothing else gets executed.

Assume some one writes an application and uses it for a board that has full hardware support for all pm features, but only a CPU idle is implemented now and now `pm_off()`. Let's also assume that application calls due to a bug `pm_off()` in a low prio thread after that thread completes. In that case, the board would just continue to run the application, but with excessive power consumption: Once the `pm_set_lowest()` returns, the board will remain in the highest power consumption mode.

The excessive power consumption is only one of the issues. If that application is build against a new release of RIOT which ships an implementation of `pm_off()` for that board, the application behavior would change: The moment the low prio thread calls `pm_off()` all other threads are never run again. The user will assume the bug is in the new RIOT release, rather than in the premature call of `pm_off()`.